mgmt/src/models/internal/natconfig/table_extend.rs

```diff
@@ -105,93 +93,138 @@ pub fn add_peering(table: &mut PerVniTable, peering: &Peering) -> Result<(), Tri
-                };
-
-                clone.remove(prefix);
-                clone_not.remove(exclude);
-                clone.insert(new_prefix);
+            // If exclusion prefix is bigger or equal to the allowed prefix, remove the allowed
+            // prefix. Given that we remove it, there's no need to compare it with the remaining
+            // exclusion prefixes.
+            if exclude.covers(prefix) {
+                result.remove(prefix);
+                break;
+            }
+
+            // If allowed prefix covers the exclusion prefix, then it means the exclusion prefix
+            // excludes a portion of this allowed prefix. We need to remove the allowed prefix, and
+            // add instead the smaller fragments resulting from the application of the exclusion
+            // prefix.
+            let Ok(mut apply_exclude_result) = apply_exclude(prefix, exclude) else {
+                return Err(());
+            };
+            if prefix.covers(exclude) {
+                result.remove(prefix);
+                result.append(&mut apply_exclude_result);
             }
         }
     }
-    (clone, clone_not)
+
+    Ok(result)
 }
 
-/// Optimize a Peering object: collapse prefixes and exclusion prefixes when possible
-fn optimize_peering(peering: &Peering) -> Peering {
-    // Collapse prefixes and exclusion prefixes
-    let mut clone = peering.clone();
-    for expose in &mut clone.local.exposes {
-        let (ips, nots) = optimize_expose(&expose.ips, &expose.nots);
-        expose.ips = ips;
-        expose.nots = nots;
+// Split a given allowed prefix into smaller allowed prefixes, taking into account exclusion
+// prefixes, to express the same range of allowed IP addresses without the need for exclusion
+// prefixes.
+fn apply_exclude(prefix: &Prefix, exclude: &Prefix) -> Result<BTreeSet<Prefix>, ()> {
+    let mut result = BTreeSet::new();
+    let mut prefix_covering_exclude = *prefix;
+    let len_diff = exclude.length() - prefix.length();
+
+    for _ in 0..len_diff {
+        let prefix_len = prefix_covering_exclude.length();
+        let prefix_address = prefix_covering_exclude.as_address();
+        let split_address = address_split(&prefix_covering_exclude);
+
+        let Ok(subprefix_low) = Prefix::try_from((prefix_address, prefix_len + 1)) else {
+            // We can't reach this: prefix_len cannot be the max length for the prefix, because it
+            // means the allowed prefix is a single IP address. If this were the case, the exclusion
+            // prefix would be a single address as well. And if this were the case, the two prefixes
+            // would be equal. This is impossible, because we returned early from the loop in the
+            // calling function if we found that exclude.covers(prefix).
+            warn!("Bug in apply_exclude logics, for subprefix_low");
+            return Err(());
```

Comment:
I think this `warn!` should be an `unreachable!` personally. Programmer error is not a recoverable thing in general

Reply:
Looks like we'll reject the config so we should return an error.